hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/RootCARotationManager.java

```diff
@@ -0,0 +1,303 @@
+    renewalGracePeriod = Duration.parse(renewalGraceDurationString);
+
+    if (checkInterval.compareTo(renewalGracePeriod) >= 0) {
+      throw new IllegalArgumentException("Property value of " +
+          HDDS_X509_CA_ROTATION_CHECK_INTERNAL +
+          " should be smaller than " + HDDS_X509_RENEW_GRACE_DURATION);
+    }
+
+    executorService = Executors.newScheduledThreadPool(1,
+        new ThreadFactoryBuilder().setNameFormat(threadName)
+            .setDaemon(true).build());
+
+    scmCertClient = scm.getScmCertificateClient();
+    scm.getSCMServiceManager().register(this);
+  }
+
+  /**
+   * Receives a notification for raft or safe mode related status changes.
+   * Stops monitor task if it's running and the current SCM becomes a
+   * follower or enter safe mode. Starts monitor if the current SCM becomes
+   * leader and not in safe mode.
+   */
+  @Override
+  public void notifyStatusChanged() {
+    // stop the monitor task
+    if (!scmContext.isLeader() || scmContext.isInSafeMode()) {
+      if (isRunning.compareAndSet(true, false)) {
+        LOG.info("notifyStatusChanged: disable monitor task.");
+      }
+      return;
+    }
+
+    if (isRunning.compareAndSet(false, true)) {
+      LOG.info("notifyStatusChanged: enable monitor task");
+    }
+    return;
+  }
+
+  /**
+   * Checks if monitor task should start (after a leader change, restart
+   * etc.) by reading persisted state.
+   * @return true if the persisted state is true, otherwise false
+   */
+  @Override
+  public boolean shouldRun() {
+    return true;
+  }
+
+  /**
+   * @return Name of this service.
+   */
+  @Override
+  public String getServiceName() {
+    return RootCARotationManager.class.getSimpleName();
+  }
+
+  /**
+   * Schedule monitor task.
+   */
+  @Override
+  public void start() throws SCMServiceException {
+    executorService.scheduleAtFixedRate(
+        new MonitorTask(scmCertClient), 0, checkInterval.toMillis(),
+        TimeUnit.MILLISECONDS);
+    LOG.info("Monitor task for root certificate {} is started with " +
+        "interval {}.", scmCertClient.getCACertificate().getSerialNumber(),
+        checkInterval);
+  }
+
+  public boolean isRunning() {
+    return isRunning.get();
+  }
+
+  /**
+   *  Task to monitor certificate lifetime and start rotation if needed.
+   */
+  public class MonitorTask implements Runnable {
+    private CertificateClient certClient;
+
+    public MonitorTask(CertificateClient client) {
+      this.certClient = client;
+    }
+
+    @Override
+    public void run() {
+      Thread.currentThread().setName(threadName +
+          (isRunning() ? "-Active" : "-Inactive"));
+      if (!isRunning.get() || isScheduled.get()) {
+        return;
+      }
+      // Lock to protect the root CA certificate rotation process,
+      // to make sure there is only one task is ongoing at one time.
+      synchronized (RootCARotationManager.class) {
+        X509Certificate rootCACert = certClient.getCACertificate();
```

Comment:
On the certClient we have a getRootCACertificate method, I believe we should use that one here.

Reply:
For ScmCertificateClient, getCACertificate returns root CA cert, getRootCACertificate returns null.

Reply:
I believe we should change this functionality some time. Let's discuss it a bit during our sync meeting. For now it's okay here.